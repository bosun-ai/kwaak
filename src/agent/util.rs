use anyhow::Context as _;
use anyhow::Result;
use swiftide_core::SimplePrompt;
use uuid::Uuid;

use crate::commands::Responder;

pub async fn rename_chat(
    query: &str,
    fast_query_provider: &dyn SimplePrompt,
    command_responder: &dyn Responder,
) -> Result<()> {
    let chat_name = fast_query_provider
        .prompt(
            format!("Give a good, short, max 60 chars title for the following query. Only respond with the title.:\n{query}")
                .into(),
        )
        .await
        .context("Could not get chat name")?
        .trim_matches('"')
        .chars()
        .take(60)
        .collect::<String>();

    command_responder.rename_chat(&chat_name);

    Ok(())
}

pub async fn create_branch_name(
    query: &str,
    uuid: &Uuid,
    fast_query_provider: &dyn SimplePrompt,
    command_responder: &dyn Responder,
) -> Result<String> {
    let name = fast_query_provider
        .prompt(
            format!("Give a good, short, max 30 chars git-branch-name for the following query. Only respond with the git-branch-name.:\n{query}")
                .into(),
        )
        .await
        .context("Could not get chat name")?
        .trim_matches('"')
        .chars()
        .take(30)
        .collect::<String>();

    // only keep ascii characters
    let name = name.chars().filter(char::is_ascii).collect::<String>();
    let name = name.to_lowercase();

    // replace all non-alphanumeric characters with dashes
    let name = name
        .chars()
        .map(|c| if c.is_alphanumeric() { c } else { '-' })
        .collect::<String>();

    // get the first 8 characters of the uuid
    let uuid_start = uuid.to_string().chars().take(8).collect::<String>();
    let branch_name = format!("kwaak/{name}-{uuid_start}");

    command_responder.rename_branch(&branch_name);

    Ok(branch_name)
}

#[cfg(test)]
mod tests {
    use swiftide_core::MockSimplePrompt;

    use crate::commands::MockResponder;
    use mockall::{predicate, PredicateBooleanExt};

    use super::*;

    #[tokio::test]
    async fn test_rename_chat() {
        let query = "This is a query";
        let mut llm_mock = MockSimplePrompt::new();
        llm_mock
            .expect_prompt()
            .returning(|_| Ok("Excellent title".to_string()));

        let mut mock_responder = MockResponder::default();

        mock_responder
            .expect_rename_chat()
            .with(predicate::eq("Excellent title"))
            .once()
            .returning(|_| ());

        rename_chat(&query, &llm_mock as &dyn SimplePrompt, &mock_responder)
            .await
            .unwrap();
    }

    #[tokio::test]
    async fn test_rename_chat_limits_60() {
        let query = "This is a query";
        let mut llm_mock = MockSimplePrompt::new();
        llm_mock
            .expect_prompt()
            .returning(|_| Ok("Excellent title".repeat(100).to_string()));

        let mut mock_responder = MockResponder::default();

        mock_responder
            .expect_rename_chat()
            .with(
                predicate::str::starts_with("Excellent title")
                    .and(predicate::function(|s: &str| s.len() == 60)),
            )
            .once()
            .returning(|_| ());

        rename_chat(&query, &llm_mock as &dyn SimplePrompt, &mock_responder)
            .await
            .unwrap();
    }

    #[tokio::test]
    async fn test_rename_branch() {
        let query = "This is a query";
        let mut llm_mock = MockSimplePrompt::new();
        llm_mock
            .expect_prompt()
            .returning(|_| Ok("excellent-name".to_string()));

        let mut mock_responder = MockResponder::default();
        let fixed_uuid = Uuid::parse_str("936DA01F9ADD4d9d80C702AF85C822A8").unwrap();

        mock_responder
            .expect_rename_branch()
            .with(predicate::str::starts_with("kwaak/excellent-name"))
            .once()
            .returning(|_| ());

        create_branch_name(
            &query,
            &fixed_uuid,
            &llm_mock as &dyn SimplePrompt,
            &mock_responder,
        )
        .await
        .unwrap();
    }

    // NOTE the prompt is intended to be limited to 30 characters, but the branch name in total
    // has 15 more characters (total 45): "kwaak/" + "-" + 8 characters from the uuid
    #[tokio::test]
    async fn test_rename_branch_limits_45() {
        let query = "This is a query";
        let mut llm_mock = MockSimplePrompt::new();
        llm_mock
            .expect_prompt()
            .returning(|_| Ok("excellent-name".repeat(100).to_string()));

        let mut mock_responder = MockResponder::default();
        let fixed_uuid = Uuid::parse_str("936DA01F9ADD4d9d80C702AF85C822A8").unwrap();

        mock_responder
            .expect_rename_branch()
            .with(
                predicate::str::starts_with("kwaak/excellent-name")
                    .and(predicate::function(|s: &str| s.len() == 45)),
            )
            .once()
            .returning(|_| ());

        create_branch_name(
            &query,
            &fixed_uuid,
            &llm_mock as &dyn SimplePrompt,
            &mock_responder,
        )
        .await
        .unwrap();
    }
}
