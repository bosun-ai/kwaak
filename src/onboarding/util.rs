use anyhow::Result;

pub fn prompt_text<'a>(prompt: &'a str, default: Option<&'a str>) -> inquire::Text<'a> {
    let mut prompt = inquire::Text::new(prompt);

    if let Some(default) = default {
        prompt = prompt.with_default(default);
    }

    prompt
}

pub fn prompt_api_key<'a>(prompt: &'a str, default: Option<&'a str>) -> inquire::Text<'a> {
    let mut prompt = inquire::Text::new(prompt).with_validator(|input: &str| {
        if input.starts_with("env:") || input.starts_with("file:") {
            Ok(inquire::validator::Validation::Valid)
        } else {
            Ok(inquire::validator::Validation::Invalid(
                "API keys must start with `env:` or `file:`".into(),
            ))
        }
    });

    if let Some(default) = default {
        prompt = prompt.with_default(default);
    }

    prompt
}

#[allow(clippy::needless_pass_by_value)]
pub fn prompt_select<T>(prompt: &str, options: Vec<T>, default: Option<T>) -> Result<String>
where
    T: std::fmt::Display + std::cmp::PartialEq + Clone,
{
    let mut prompt = inquire::Select::new(prompt, options.clone());

    if let Some(default) = default {
        debug_assert!(
            options.contains(&default),
            "{} is not in the list of options, valid: {}",
            default,
            options
                .iter()
                .map(ToString::to_string)
                .collect::<Vec<_>>()
                .join(", ")
        );
        if let Some(idx) = options.iter().position(|l| l == &default) {
            prompt = prompt.with_starting_cursor(idx);
        }
    }

    Ok(prompt.prompt()?.to_string())
}
