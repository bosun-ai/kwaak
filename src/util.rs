use swiftide::traits::{CommandError, CommandOutput};

pub fn strip_markdown_tags(text: &str) -> String {
    if text.starts_with("```markdown") && text.ends_with("```") {
        text[12..text.len() - 3].trim().to_string()
    } else {
        text.to_string()
    }
}

// TODO: Would be nice if this was a method on a custom result in swiftide
pub fn accept_non_zero_exit(
    result: Result<CommandOutput, CommandError>,
) -> Result<CommandOutput, CommandError> {
    match result {
        Ok(output) | Err(CommandError::NonZeroExit(output)) => Ok(output),
        Err(err) => Err(err),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_strip_markdown_tags() {
        // Case: String with surrounding ```markdown tags
        let input = "```markdown\nThis is a test.\n```";
        let expected = "This is a test.";
        assert_eq!(strip_markdown_tags(input), expected);

        // Case: String without surrounding ```markdown tags
        let input = "This is a test.";
        let expected = "This is a test.";
        assert_eq!(strip_markdown_tags(input), expected);

        // Case: Empty string
        let input = "";
        let expected = "";
        assert_eq!(strip_markdown_tags(input), expected);

        // Case: String with only opening ```markdown tag
        let input = "```markdown\nThis is a test.";
        let expected = "```markdown\nThis is a test.";
        assert_eq!(strip_markdown_tags(input), expected);

        // Case: String with only closing ``` tag
        let input = "This is a test.\n```";
        let expected = "This is a test.\n```";
        assert_eq!(strip_markdown_tags(input), expected);

        // Case: Raw string
        let input = r"```markdown
        This is a test.
        ```";
        let expected = "This is a test.";
        assert_eq!(strip_markdown_tags(input), expected);
    }
}
