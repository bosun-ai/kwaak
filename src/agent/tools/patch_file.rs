use std::{borrow::Cow, str::FromStr};

use anyhow::{Context as _, Result};
use swiftide::{
    chat_completion::{errors::ToolError, ToolOutput},
    traits::{AgentContext, Command},
};
use swiftide_macros::tool;

use crate::util::accept_non_zero_exit;

// TODO:
// - Fix hunk header parsing
// - Check if line numbers match?
// - Handle ambigious

const REPLACE_PATCH_DESCRIPTION: &str = "Replace content with a Unified format git patch

Here is an example of a Unified format git patch:

```patch
--- a/src/evaluations/patch.rs
+++ b/src/evaluations/patch.rs
@@ -43,7 +43,7 @@ fn prompt() -> String {
             self._content_consumed = True
         ```
 
-        Apply only these fixes, do not make any other changes to the code. The file is long and the modifications are small.
+        Apply only these fixes, do not make any other changes to the code. The file is long and the modifications are small. Start by reading the file.
     \"}.to_string()
 }
 
-- 
```
";

#[tool(
    description = REPLACE_PATCH_DESCRIPTION,
    param(name = "file_name", description = "Full path of the file"),
    param(name = "patch", description = "Unified format git patch to apply"),
)]
async fn patch_file(
    context: &dyn AgentContext,
    file_name: &str,
    patch: &str,
) -> Result<ToolOutput, ToolError> {
    let cmd = Command::ReadFile(file_name.into());
    let old_content = accept_non_zero_exit(context.exec_cmd(&cmd).await)?.output;

    // let patch = fix_hunk_headers(&content, &patch);

    let patch = match diffy::Patch::from_str(patch) {
        Ok(patch) => patch,
        Err(err) => {
            return Ok(ToolOutput::Fail(format!("Failed to parse patch: {err}")));
        }
    };
    let patched = diffy::apply(&old_content, &patch).context("Failed to apply patch")?;

    let cmd = Command::WriteFile(file_name.into(), patched);
    context.exec_cmd(&cmd).await?;

    Ok(ToolOutput::Text("Patch applied successfully".into()))
}

// llms are dumb and cannot count
//
// However, with a patch we can reasonably fix the headers
// by searching in the neighboring lines of the original hunk header
fn find_candidates<'a>(content: &str, hunks: &'a [Hunk]) -> Vec<Candidate<'a>> {
    let mut candidates = Vec::new();

    for (line_n, line) in content.lines().enumerate() {
        // 1. Check if a hunk matches the line, then create a candidate if it does
        if let Some(hunk) = hunks.iter().find(|h| h.matches(line, 0, false)) {
            tracing::warn!(line, "Found hunk match; creating new candidate");
            candidates.push(Candidate::new(line_n, hunk));
        }

        // 2. For each active candidate, check if the next line matches. If it does, increment the
        // the index of the candidate. Otherwise, remove the candidate
        candidates.retain_mut(|c| {
            if c.is_complete() {
                tracing::warn!("Candidate already completed");
                true
            } else if c.next_line_matches(line) {
                tracing::warn!(line, "Candidate matched line");
                c.current_line += 1;
                true
            } else {
                tracing::warn!(line, "Removing candidate");
                false
            }
        });
    }

    candidates
}

/// Takes a list of candidates and rebuits the hunk headers
fn rebuild_hunks(candidates: &[Candidate<'_>]) -> Vec<Hunk> {
    // Assume that the candidates are sorted by the start line
    // Then we can just iterate over the candidates and update the ranges
    //
    // TODO: Deal with duplicated candidates

    let mut current_offset: isize = 0;
    let mut hunks = Vec::new();

    for candidate in candidates {
        let source_header = candidate.updated_source_header();

        let dest_header = candidate.updated_dest_header(current_offset);
        current_offset += candidate.offset();

        let mut hunk = candidate.hunk.clone();
        hunk.header.fixed_source_range = Some(source_header);
        hunk.header.fixed_dest_range = Some(dest_header);
        hunks.push(hunk);
    }

    hunks
}

/// Splits the patch into a tuple of the hunk header with the full hunk (including the header)
fn parse_hunks(patch: &str) -> Result<Vec<Hunk>> {
    let mut hunks = Vec::new();
    let mut current_hunk_lines = Vec::new();

    for line in patch.lines() {
        if line.starts_with("@@") {
            if !current_hunk_lines.is_empty() {
                let hunk = Hunk::from_str(&current_hunk_lines.join("\n"))?;
                hunks.push(hunk);
            }

            current_hunk_lines = vec![line];
        } else if !current_hunk_lines.is_empty() {
            current_hunk_lines.push(line);
        }
    }

    if !current_hunk_lines.is_empty() {
        let hunk = Hunk::from_str(&current_hunk_lines.join("\n"))?;
        hunks.push(hunk);
    }

    Ok(hunks)
}

#[derive(Clone, Debug)]
struct HeaderRange {
    /// The line number the patch starts at
    start: usize,
    /// The line numbers visible for the patch
    range: usize,
}

#[derive(Clone, Debug)]
struct HunkHeader {
    source_range: HeaderRange,
    dest_range: HeaderRange,

    // Optional values after fixing the ranges
    fixed_source_range: Option<HeaderRange>,
    fixed_dest_range: Option<HeaderRange>,
}

#[derive(Clone, Debug, strum_macros::EnumIs)]
enum HunkLine {
    Context(String),
    Added(String),
    Removed(String),
}

impl HunkLine {
    pub fn content(&self) -> &str {
        match self {
            HunkLine::Context(s) => s,
            HunkLine::Added(s) => s,
            HunkLine::Removed(s) => s,
        }
    }

    pub fn as_patch_line(&self) -> Cow<str> {
        match self {
            HunkLine::Context(s) => Cow::Borrowed(s),
            HunkLine::Added(s) => Cow::Owned(format!("+{}", s)),
            HunkLine::Removed(s) => Cow::Owned(format!("-{}", s)),
        }
    }
}

#[derive(Clone, Debug)]
struct Hunk {
    /// The parsed header of the hunk
    header: HunkHeader,

    /// Parsed lines of the hunk
    lines: Vec<HunkLine>,

    /// The full hunk body
    body: String,
}

impl Hunk {
    fn matchable_lines(&self) -> impl Iterator<Item = &HunkLine> {
        self.lines
            .iter()
            .filter(|l| l.is_removed() || l.is_context())
    }

    pub fn matches(&self, line: &str, index: usize, log: bool) -> bool {
        let expected = self
            .matchable_lines()
            .skip(index)
            .map(HunkLine::content)
            .next();

        let outcome = expected.map(str::trim) == Some(line.trim());

        if log {
            if outcome {
                tracing::warn!(line, expected, "Matched line");
            } else {
                tracing::debug!(line, expected, "Did not match line");
            }
        }
        outcome
    }

    pub fn render_updated(&self) -> Result<String> {
        // Extract any context after the second @@ block to add to the new header line
        // i.e. with `@@ -1,2 +2,1 @@ my_function()` we want my_function() to be included
        let header_context = self
            .body
            .lines()
            .next()
            .unwrap_or_default()
            .rsplit("@@")
            .next()
            .unwrap_or_default();

        let source = self
            .header
            .fixed_source_range
            .as_ref()
            .context("Expected")?;
        let dest = self.header.fixed_dest_range.as_ref().context("Expected")?;

        let mut updated = format!(
            "@@ -{},{} +{},{} @@{header_context}\n",
            source.start + 1,
            source.range,
            dest.start + 1,
            dest.range
        );

        for line in &self.lines {
            updated.push_str(&line.as_patch_line());
            updated.push('\n');
        }

        Ok(updated)
    }
}

/// A hunk that is found in a file
#[derive(Clone, Debug)]
struct Candidate<'a> {
    /// The line number in the file we started at
    start: usize,

    /// The current line we are matchin against
    current_line: usize,

    hunk: &'a Hunk,
}

impl<'a> Candidate<'a> {
    pub fn new(line: usize, hunk: &'a Hunk) -> Self {
        Self {
            start: line,
            current_line: 0,
            hunk,
        }
    }

    /// Number difference in visible lines between the source and destination for the next hunk
    ///
    /// If lines were added, the following hunk will start at an increased line number, if lines
    /// were removed, the following hunk will start at a decreased line number
    pub fn offset(&self) -> isize {
        self.hunk.lines.iter().filter(|l| l.is_added()).count() as isize
            - self.hunk.lines.iter().filter(|l| l.is_removed()).count() as isize
    }

    pub fn next_line_matches(&self, line: &str) -> bool {
        self.hunk.matches(line, self.current_line, true)
    }

    pub fn is_complete(&self) -> bool {
        // We increment one over the current line, so if we are at the end of the hunk, we are done
        self.current_line == self.hunk.matchable_lines().count()
    }

    pub fn updated_source_header(&self) -> HeaderRange {
        let source_lines = self
            .hunk
            .lines
            .iter()
            .filter(|l| l.is_removed() || l.is_context())
            .count();

        let source_start = self.start;

        HeaderRange {
            start: source_start,
            range: source_lines,
        }
    }

    pub fn updated_dest_header(&self, offset: isize) -> HeaderRange {
        let dest_lines = self
            .hunk
            .lines
            .iter()
            .filter(|l| l.is_added() || l.is_context())
            .count();

        // The offset is the sum off removed and added lines by preceding hunks
        let dest_start = self.start.saturating_add_signed(offset);

        HeaderRange {
            start: dest_start,
            range: dest_lines,
        }
    }
}

impl FromStr for Hunk {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let header: HunkHeader = s.parse()?;
        let lines = s
            .lines()
            .skip(1)
            .map(FromStr::from_str)
            .collect::<Result<Vec<HunkLine>>>()?;

        Ok(Hunk {
            header,
            lines,
            body: s.into(),
        })
    }
}

impl FromStr for HunkLine {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if let Some(line) = s.strip_prefix('+') {
            Ok(HunkLine::Added(line.into()))
        } else if let Some(line) = s.strip_prefix('-') {
            Ok(HunkLine::Removed(line.into()))
        } else {
            Ok(HunkLine::Context(s.into()))
        }
    }
}

// For the header we just parse, as there is nothing to borrow
impl std::str::FromStr for HunkHeader {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if !s.starts_with("@@") {
            anyhow::bail!("Hunk header must start with @@");
        }

        let parts: Vec<&str> = s.split_whitespace().collect();
        if parts.len() < 4 {
            anyhow::bail!("Invalid hunk header format");
        }

        let old_range = parts[1].split(',').collect::<Vec<&str>>();
        let new_range = parts[2].split(',').collect::<Vec<&str>>();

        if old_range.len() != 2 || new_range.len() != 2 {
            anyhow::bail!("Invalid range format in hunk header");
        }

        let old_lines = HeaderRange {
            start: old_range[0]
                .replace('-', "")
                .parse()
                .context("Invalid old start line")?,
            range: old_range[1].parse().context("Invalid old range")?,
        };

        let new_lines = HeaderRange {
            start: new_range[0]
                .replace('+', "")
                .parse()
                .context("Invalid new start line")?,
            range: new_range[1].parse().context("Invalid new range")?,
        };

        Ok(HunkHeader {
            source_range: old_lines,
            dest_range: new_lines,
            fixed_source_range: None,
            fixed_dest_range: None,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    const BAD_SINGLE_HUNK: &str = indoc::indoc! {"--- a/src/evaluations/fixtures/swebench_2148/models.py
+++ b/src/evaluations/fixtures/swebench_2148/models.py
@@ -637,6 +637,7 @@ def iter_content(self, chunk_size=1, decode_unicode=False):
                 except IncompleteRead as e:
                     raise ChunkedEncodingError(e)
                 except DecodeError as e:
                     raise ContentDecodingError(e)
+                except socket.error as e:
+                    raise ConnectionError(e)
             except AttributeError:
                 # Standard file-like object.
                 while True:
"};
    const BAD_PATCH: &str = indoc::indoc! {"--- a/src/evaluations/fixtures/swebench_2148/models.py
+++ b/src/evaluations/fixtures/swebench_2148/models.py
@@ -637,6 +637,7 @@ def iter_content(self, chunk_size=1, decode_unicode=False):
                 except IncompleteRead as e:
                     raise ChunkedEncodingError(e)
                 except DecodeError as e:
                     raise ContentDecodingError(e)
+                except socket.error as e:
+                    raise ConnectionError(e)
             except AttributeError:
                 # Standard file-like object.
                 while True:
@@ -652,8 +653,9 @@ def iter_content(self, chunk_size=1, decode_unicode=False):
                     yield chunk
 
-            self._content_consumed = True
+            
+            
 
+        
         # simulate reading small chunks of the content
         reused_chunks = iter_slices(self._content, chunk_size)
         
@@ -664,6 +666,8 @@ def iter_content(self, chunk_size=1, decode_unicode=False):
 
         if decode_unicode:
             chunks = stream_decode_response_unicode(chunks, self)
+
+        finally:
+            self._content_consumed = True
 
         return chunks


"};

    #[test]
    fn test_split_patch_into_hunks() {
        // TODO: Add some more unit tests to check parsing is correct
        let hunks = parse_hunks(BAD_PATCH).unwrap();
        assert_eq!(hunks.len(), 3);

        let header = &hunks[0].header;

        assert_eq!(header.source_range.start, 637);
        assert_eq!(header.source_range.range, 6);

        assert_eq!(header.dest_range.start, 637);
        assert_eq!(header.dest_range.range, 7);

        let header = &hunks[1].header;
        assert_eq!(header.source_range.start, 652);
        assert_eq!(header.source_range.range, 8);

        assert_eq!(header.dest_range.start, 653);
        assert_eq!(header.dest_range.range, 9);

        let header = &hunks[2].header;

        assert_eq!(header.source_range.start, 664);
        assert_eq!(header.source_range.range, 6);

        assert_eq!(header.dest_range.start, 666);
        assert_eq!(header.dest_range.range, 8);
    }

    #[test_log::test]
    fn test_find_candidates_single_hunk() {
        let hunks = parse_hunks(&BAD_SINGLE_HUNK).unwrap();
        assert_eq!(hunks.len(), 1);
        let content = std::fs::read_to_string("src/evaluations/fixtures/swebench_2148/models.py")
            .expect("Failed to read file");
        let candidates = find_candidates(&content, &hunks);
        dbg!(&candidates);
        assert_eq!(candidates.len(), 1);

        let hunk = rebuild_hunks(&candidates).first().unwrap().clone();

        dbg!(&hunk);

        assert_eq!(hunk.header.fixed_source_range.as_ref().unwrap().start, 641); // One less than
                                                                                 // in the source file
        assert_eq!(hunk.header.fixed_source_range.as_ref().unwrap().range, 7);
        assert_eq!(hunk.header.fixed_dest_range.as_ref().unwrap().start, 641);
        assert_eq!(hunk.header.fixed_dest_range.as_ref().unwrap().range, 9);
        assert_eq!(candidates.first().unwrap().offset(), 2);

        insta::assert_snapshot!(hunk.render_updated().unwrap());
    }

    #[test_log::test]
    fn test_find_candidates_multiple_hunks() {
        let hunks = parse_hunks(&BAD_PATCH).unwrap();
        let content = std::fs::read_to_string("src/evaluations/fixtures/swebench_2148/models.py")
            .expect("Failed to read file");

        let adjusted_hunks = find_candidates(&content, &hunks);
        assert_eq!(adjusted_hunks.len(), hunks.len());
    }

    // #[test]
    // fn test_fix_hunk_header() {
    //     diffy::Patch::from_str(BAD_PATCH).unwrap();
    // }
}
