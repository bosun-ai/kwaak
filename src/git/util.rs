use std::path::Path;

use anyhow::Result;
use swiftide_core::{Command, ToolExecutor};

use crate::util::accept_non_zero_exit;

/// Get the diff from a tool executor
pub async fn diff(executor: &dyn ToolExecutor, base_sha: &str, color: bool) -> Result<String> {
    let color = if color { "--color=always" } else { "" };
    let cmd = Command::shell(format!("git diff {color} {base_sha}",));

    let mut output = accept_non_zero_exit(executor.exec_cmd(&cmd).await)?.output;

    if output.is_empty() {
        output = "No changes".to_string();
    }

    Ok(output)
}

/// Gets the main branch sync
///
/// WARN: Not to be used outside init and tests
pub fn main_branch(workdir: impl AsRef<Path>) -> String {
    const DEFAULT_BRANCH: &str = "main";
    // Tries to get it from the ref if present, otherwise sets it, the uses rev-parse to get the branch
    // The ref can be missing if the repo was never cloned (i.e. author and pushed to github directly)
    const MAIN_BRANCH_CMD: &str =
    "(git symbolic-ref refs/remotes/origin/HEAD >/dev/null 2>&1 || git remote set-head origin --auto >/dev/null 2>&1) && git rev-parse --abbrev-ref origin/HEAD";

    let Ok(output) = std::process::Command::new("sh")
        .arg("-c")
        .arg(MAIN_BRANCH_CMD)
        .current_dir(workdir.as_ref())
        .output()
    else {
        return DEFAULT_BRANCH.to_string();
    };

    let parsed = std::str::from_utf8(&output.stdout)
        .unwrap_or(DEFAULT_BRANCH)
        .trim_start_matches("origin/")
        .trim();

    if parsed.is_empty() {
        DEFAULT_BRANCH.to_string()
    } else {
        parsed.to_string()
    }
}

/// Checks if the current branch is dirty or contains untracked files
///
/// If for some reason the command fails, it will return true
pub async fn is_dirty(workdir: impl AsRef<Path>) -> bool {
    tokio::process::Command::new("git")
        .arg("diff-index")
        .arg("--quiet")
        .arg("HEAD")
        .current_dir(workdir.as_ref())
        .output()
        .await
        .map(|output| !output.status.success())
        .unwrap_or(true)
}
