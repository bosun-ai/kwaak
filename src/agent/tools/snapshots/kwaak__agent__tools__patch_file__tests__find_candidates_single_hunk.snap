---
source: src/agent/tools/patch_file.rs
expression: hunk.render_updated().unwrap()
---
@@ -642,7 +642,9 @@ def iter_content(self, chunk_size=1, decode_unicode=False):
                 except IncompleteRead as e:
                     raise ChunkedEncodingError(e)
                 except DecodeError as e:
                     raise ContentDecodingError(e)
+                except socket.error as e:
+                    raise ConnectionError(e)
             except AttributeError:
                 # Standard file-like object.
                 while True:
